projects/kit/components/slider/slider-key-steps.directive.ts

```diff
@@ -0,0 +1,129 @@
+        return isNativeFocused(this.nativeFocusableElement);
+    }
+
+    constructor(
+        @Optional()
+        @Self()
+        @Inject(NgControl)
+        control: NgControl | null,
+        @Inject(ChangeDetectorRef) changeDetectorRef: ChangeDetectorRef,
+        @Inject(ElementRef) private readonly elementRef: ElementRef<HTMLInputElement>,
+        @Inject(TuiSliderComponent) private readonly slider: TuiSliderComponent,
+    ) {
+        super(control, changeDetectorRef);
+    }
+
+    @HostListener('change')
+    onInput() {
+        const {valuePercentage} = this.slider;
+        const [lowerStep, upperStep] = findKeyStepsBoundariesByFn(
+            this.keySteps,
+            ([keyStepPercentage, _]) => valuePercentage <= keyStepPercentage,
+        );
+        const newControlValue = transformToControlValue(
+            valuePercentage,
+            lowerStep,
+            upperStep,
+        );
+
+        this.updateValue(newControlValue);
+    }
+
+    writeValue(controlValue: number | null) {
+        if (controlValue === null) {
+            return;
+        }
+
+        const [lowerStep, upperStep] = findKeyStepsBoundariesByFn(
+            this.keySteps,
+            ([_, keyStepValue]) => controlValue <= keyStepValue,
+        );
+
+        this.slider.value = this.transformToNativeValue(
+            controlValue,
+            lowerStep,
+            upperStep,
+        );
+    }
+
+    protected getFallbackValue(): number {
+        return 0;
+    }
+
+    private transformToNativeValue(
+        controlValue: number,
+        [upperStepPercent, upperStepValue]: [number, number],
+        [lowerStepPercent, lowerStepValue]: [number, number],
+    ): number {
+        const {min, max} = this.slider;
+        const ratio = (controlValue - lowerStepValue) / (upperStepValue - lowerStepValue);
+        const newValuePercentage =
+            (upperStepPercent - lowerStepPercent) * ratio + lowerStepPercent;
+
+        return (newValuePercentage * (max - min)) / 100;
+    }
+}
+
+function checkHasMinMaxPercents(steps: TuiKeySteps): boolean {
+    return !steps.length || (steps[0][0] === 0 && steps[steps.length - 1][0] === 100);
+}
+
+function findKeyStepsBoundariesByFn(
+    keySteps: TuiKeySteps,
+    fn: ([keyStepPercent, keyStepValue]: [number, number]) => boolean,
+): [[number, number], [number, number]] {
+    const keyStepUpperIndex = keySteps.findIndex((ketStep, i) => i && fn(ketStep));
+    const lowerStep = keySteps[keyStepUpperIndex - 1];
+    const upperStep = keySteps[keyStepUpperIndex];
+
+    return [lowerStep, upperStep];
+}
+
+function transformToControlValue(
+    valuePercentage: number,
+    [upperStepPercent, upperStepValue]: [number, number],
+    [lowerStepPercent, lowerStepValue]: [number, number],
+): number {
+    const ratio =
+        (valuePercentage - lowerStepPercent) / (upperStepPercent - lowerStepPercent);
+
+    return (upperStepValue - lowerStepValue) * ratio + lowerStepValue;
+}
```

Comment:
nit: move to `slider-key-steps.utils.ts`

Reply:
✅

Reply:
Let's keep them here so they are not exported and do not pollute public API

Reply:
but they are not exported to index.ts